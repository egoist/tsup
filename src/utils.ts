import fs from 'fs'
import JoyCon from 'joycon'
import stripJsonComments from 'strip-json-comments'
import resovleFrom from 'resolve-from'

const joycon = new JoyCon()

joycon.addLoader({
  test: /\.json$/,
  async load(filepath) {
    const content = stripJsonComments(
      await fs.promises.readFile(filepath, 'utf8')
    )
    return JSON.parse(content)
  },
})

// No backslash in path
function slash(input: string) {
  return input.replace(/\\/g, '/')
}

export type External =
  | string
  | RegExp
  | ((id: string, parentId?: string) => boolean)

export function isExternal(
  externals: External | External[],
  id: string,
  parentId?: string
) {
  id = slash(id)

  if (!Array.isArray(externals)) {
    externals = [externals]
  }

  for (const external of externals) {
    if (
      typeof external === 'string' &&
      (id === external || id.includes(`/node_modules/${external}/`))
    ) {
      return true
    }
    if (external instanceof RegExp) {
      if (external.test(id)) {
        return true
      }
    }
    if (typeof external === 'function') {
      if (external(id, parentId)) {
        return true
      }
    }
  }

  return false
}

export function loadTsConfig(cwd: string) {
  return joycon.load(['tsconfig.build.json', 'tsconfig.json'], cwd)
}

export async function getDeps(cwd: string) {
  const data = await loadPkg(cwd)

  const deps = Array.from(new Set([
    ...Object.keys(data.dependencies || {}),
    ...Object.keys(data.peerDependencies || {})
  ]))

  return deps
}

export async function loadPkg(cwd: string) {
  const { data } = await joycon.load(['package.json'], cwd)
  return data || {}
}

export function getBabel(): null | typeof import('@babel/core') {
  const p = resovleFrom.silent(process.cwd(), '@babel/core')
  return p && require(p)
}
