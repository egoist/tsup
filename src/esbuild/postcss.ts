import fs from 'node:fs'
import { type Loader, type Plugin, transform } from 'esbuild'
import { getPostcss } from '../utils'
import type { Result } from 'postcss-load-config'

export const postcssPlugin = ({
  css,
  inject,
  cssLoader,
  fileExtensions = ["css"]
}: {
  css?: Map<string, string>
  inject?: boolean | ((css: string, fileId: string) => string | Promise<string>)
  cssLoader?: Loader,
  fileExtensions?: string[]
}): Plugin => {
  return {
    name: 'postcss',

    setup(build) {
      let configCache: Result

      const getPostcssConfig = async () => {
        const loadConfig = require('postcss-load-config')

        if (configCache) {
          return configCache
        }

        try {
          const result = await loadConfig({}, process.cwd())
          configCache = result
          return result
        } catch (error: any) {
          if (error.message.includes('No PostCSS Config found in')) {
            const result = { plugins: [], options: {} }
            return result
          }
          throw error
        }
      }

      build.onResolve({ filter: /^#style-inject$/ }, () => {
        return { path: '#style-inject', namespace: '#style-inject' }
      })

      build.onLoad(
        { filter: /^#style-inject$/, namespace: '#style-inject' },
        () => {
          return {
            // Taken from https://github.com/egoist/style-inject/blob/master/src/index.js (MIT)
            contents: `
          export default function styleInject(css, { insertAt } = {}) {
            if (!css || typeof document === 'undefined') return
          
            const head = document.head || document.getElementsByTagName('head')[0]
            const style = document.createElement('style')
            style.type = 'text/css'
          
            if (insertAt === 'top') {
              if (head.firstChild) {
                head.insertBefore(style, head.firstChild)
              } else {
                head.appendChild(style)
              }
            } else {
              head.appendChild(style)
            }
          
            if (style.styleSheet) {
              style.styleSheet.cssText = css
            } else {
              style.appendChild(document.createTextNode(css))
            }
          }
          `,
            loader: 'js',
          }
        },
      )

      build.onLoad({ filter: new RegExp(fileExtensions.map(ext => `\\.${ext}$`).join("|")) }, async (args) => {
        let contents: string

        if (css && args.path.endsWith('.svelte.css')) {
          contents = css.get(args.path)!
        } else {
          contents = await fs.promises.readFile(args.path, 'utf8')
        }

        // Load postcss config
        const { plugins, options } = await getPostcssConfig()

        if (plugins && plugins.length > 0) {
          // Load postcss
          const postcss = getPostcss()
          if (!postcss) {
            return {
              errors: [
                {
                  text: `postcss is not installed`,
                },
              ],
            }
          }

          // Transform CSS
          const result = await postcss
            ?.default(plugins)
            .process(contents, { ...options, from: args.path })

          contents = result.css
        }

        if (inject) {
          contents = (
            await transform(contents, {
              minify: build.initialOptions.minify,
              minifyIdentifiers: build.initialOptions.minifyIdentifiers,
              minifySyntax: build.initialOptions.minifySyntax,
              minifyWhitespace: build.initialOptions.minifyWhitespace,
              logLevel: build.initialOptions.logLevel,
              loader: 'css',
            })
          ).code

          contents =
            typeof inject === 'function'
              ? await inject(JSON.stringify(contents), args.path)
              : `import styleInject from '#style-inject';styleInject(${JSON.stringify(
                  contents,
                )})`

          return {
            contents,
            loader: 'js',
          }
        }

        return {
          contents,
          loader: cssLoader ?? 'css',
        }
      })
    },
  }
}
