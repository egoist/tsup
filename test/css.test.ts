import { expect, test } from 'vitest'
import { getTestName, run } from './utils'

test('import css', async () => {
  const { output, outFiles, getFileContent } = await run(getTestName(), {
    'input.ts': `
    import './foo.css'
    `,
    'postcss.config.js': `
    module.exports = {
      plugins: [require('postcss-simple-vars')()]
    }
    `,
    'foo.css': `
  $color: blue;

  .foo {
    color: $color;
  }
    `,
  })

  expect(output, `""`).toMatchSnapshot()
  expect(await getFileContent("dist/input.css")).not.toContain("$color")
  expect(outFiles).toEqual(['input.css', 'input.js'])
})

test('support alternative postcss file extension', async () => {
  const { output, outFiles, getFileContent } = await run(getTestName(), {
    'input.ts': `
    import './foo.postcss'
    `,
    'postcss.config.js': `
    module.exports = {
      plugins: [require('postcss-simple-vars')()]
    }
    `,
    'foo.postcss': `
  $color: blue;

  .foo {
    color: $color;
  }
    `,
    'tsup.config.ts': `
      export default {
        loader: {
          '.postcss': 'css'
        },
        postcssFileExtensions: ['postcss']
      }
      `,
  })
  

  expect(output, `""`).toMatchSnapshot()
  expect(await getFileContent("dist/input.css")).not.toContain("$color")
  expect(outFiles).toEqual(['input.css', 'input.js'])
})

test('support tailwindcss postcss plugin', async () => {
  const { output, outFiles } = await run(getTestName(), {
    'input.ts': `
      import './foo.css'
    `,
    'postcss.config.js': `
      module.exports = {
        plugins: {
          tailwindcss: {},
          autoprefixer: {},
        }
      }
    `,
    'foo.css': `
      @tailwind base;
      @tailwind components;
      @tailwind utilities;
    `,
  })
  expect(output).toMatchSnapshot()
  expect(outFiles).toEqual(['input.css', 'input.js'])
})

test('import css in --dts', async () => {
  const { output, outFiles } = await run(
    getTestName(),
    {
      'input.ts': `
    import './foo.css'
    `,
      'foo.css': `
  .foo {
    color: blue
  }
    `,
    },
    { flags: ['--dts'] },
  )

  expect(output).toMatchSnapshot()
  expect(outFiles).toEqual(['input.css', 'input.d.ts', 'input.js'])
})
