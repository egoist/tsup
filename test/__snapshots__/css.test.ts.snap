// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`import css 1`] = `
""use strict";
"
`;

exports[`import css in --dts 1`] = `
""use strict";
"
`;

exports[`support alternative postcss file extension 1`] = `
""use strict";
"
`;

exports[`support tailwindcss postcss plugin 1`] = `
""use strict";
"
`;
